extensions/rocksdb-repos/DatabaseContentRepository.cpp

```diff
@@ -198,27 +212,63 @@ bool DatabaseContentRepository::exists(const minifi::ResourceClaim &streamId) {
 }
 
 bool DatabaseContentRepository::removeKey(const std::string& content_path) {
-  if (!is_valid_ || !db_) {
-    logger_->log_error("DB is not valid, could not delete %s", content_path);
-    return false;
-  }
-  auto opendb = db_->open();
-  if (!opendb) {
-    logger_->log_error("Could not open DB, did not delete %s", content_path);
-    return false;
-  }
-  rocksdb::Status status;
-  status = opendb->Delete(rocksdb::WriteOptions(), content_path);
-  if (status.ok()) {
-    logger_->log_debug("Deleting resource %s", content_path);
-    return true;
-  } else if (status.IsNotFound()) {
-    logger_->log_debug("Resource %s was not found", content_path);
-    return true;
-  } else {
-    logger_->log_error("Attempted, but could not delete %s", content_path);
-    return false;
+  if (purge_period_ == std::chrono::seconds(0)) {
+    if (!is_valid_ || !db_)
+      return false;
+    // synchronous deletion
+    auto opendb = db_->open();
+    if (!opendb) {
+      return false;
+    }
+    rocksdb::Status status = opendb->Delete(rocksdb::WriteOptions(), content_path);
+    if (status.ok()) {
+      logger_->log_debug("Deleting resource %s", content_path);
+      return true;
+    } else if (status.IsNotFound()) {
+      logger_->log_debug("Resource %s was not found", content_path);
+      return true;
+    } else {
+      logger_->log_debug("Attempted, but could not delete %s", content_path);
+      return false;
+    }
   }
+  // asynchronous deletion
+  std::lock_guard guard(keys_mtx_);
+  logger_->log_debug("Staging resource for deletion %s", content_path);
+  keys_to_delete_.push_back(content_path);
+  return true;
+}
+
+void DatabaseContentRepository::runGc() {
+  do {
+    auto opendb = db_->open();
+    if (!opendb) {
+      continue;
+    }
+    std::vector<std::string> keys;
+    {
+      std::lock_guard guard(keys_mtx_);
+      keys = std::exchange(keys_to_delete_, std::vector<std::string>{});
+    }
+    auto batch = opendb->createWriteBatch();
+    for (auto& key : keys) {
+      batch.Delete(key);
+    }
+    rocksdb::Status status;
+    status = opendb->Write(rocksdb::WriteOptions(), &batch);
+    if (status.ok()) {
```

Comment:
We should handle the case when there is an interruption (kill, shutdown, etc) between adding the keys to the vector, and actually deleting them, including an interruption between these lines: emptying `keys_to_delete_` and actually deleting. For the latter part, doing the delete before emptying the list, but while holding the mutex may do the trick, but may not be the best idea from a performance standpoint. What do you think?

Do we store `keys_to_delete_` on disk?

Reply:
on a restart we first comb through the content repository and delete all content without a flowfile referencing them (in `clearOrphans`) that should prevent leaking content

Reply:
added